### PR TITLE
bcm53xx: enhance support for Linksys Panamera aka EA9500 v1 or v1.1

### PR DESCRIPTION
**Pull request by npcomplete13**
Based on my accepted patches, we can now enable image generation for EA9500 router.

This pull request consists of:

1. Backporting accepted patches for EA9500 and BCM5301X platform
2. Add LED config, Network config and partialboot flag related patches
3. Enable image generation itself

This has been tested locally by me and several users [here](https://forum.openwrt.org/t/build-for-linksys-ea9500/1817)

**Hardware Info:**

- Processor	- Broadcom BCM4709C0KFEBG dual-core @ 1.4 GHz
- Switch		- BCM53012 in BCM4709C0KFEBG & external BCM53125
- DDR3 RAM	- 256 MB
- Flash		- 128 MB (Toshiba TC58BVG0S3HTA00)
- 2.4GHz		- BCM4366 4×4 2.4/5G single chip 802.11ac SoC
- Power Amp	- Skyworks SE2623L 2.4 GHz power amp (x4)
- 5GHz x 2	- BCM4366 4×4 2.4/5G single chip 802.11ac SoC
- Power Amp	- PLX Technology PEX8603 3-lane, 3-port PCIe switch
- Ports		- 8 Ports, 1 WAN Ports
- Antennas	- 8 Antennas
- Serial Port	- @j6 [GND,TX,RX] (VCC NC)    115200 8n1

**Flashing Instructions:**

1. Connect a USB-TTL table to J6 on the router as well as a ethernet cable to a lan port and your PC.  
2. Power-on the router. 
3. Use putty or a serial port program to view the terminal. Hit Ctrl+C and interrupt the CFE terminal terminal. 
4. Setup a TFTP server on your local machine at setup you local IP to 192.168.1.2
5. Start the TFTP Server
6. Run following commands at the CFE terminal
```
flash -noheader 192.168.1.2:/openwrt.trx nflash0.trx
flash -noheader 192.168.1.2:/openwrt.trx nflash0.trx2
nvram set bootpartition=0 && nvram set partialboots=0 && nvram commit
```
7. Reboot router to be presented by OpenWrt

Note: Only installation method via serial cable is supported at the moment. The trx firmware has to be flashed to both the partitions using following commands from CFE prompt. This will cover US and Non-US variants.

Signed-off-by: Vivek Unune <npcomplete13@gmail.com>

### CODE REVIEW

File: target/linux/bcm53xx/base-files/etc/board.d/02_network [hunk 1]
```diff
@@ -31,6 +31,10 @@ bcm53xx_setup_interfaces()
 		ucidef_add_switch "switch0" \
 			"0:lan" "1:lan" "2:lan" "3:lan" "4:wan" "5t@eth0"
 		;;
+	linksys,panamera)
+		ucidef_add_switch "switch1" \
+			"0:lan" "1:lan:7" "2:lan:4" "3:lan:8" "4:wan" "5t@eth0"
```
Comment (adschm) [inline target/linux/bcm53xx/base-files/etc/board.d/02_network:36]: What's the :7, :4; :8 supposed to achieve? Are these number actually used in Luci?

File: target/linux/bcm53xx/base-files/etc/init.d/clear_partialboot [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+START=97
+boot() {
+. /lib/functions.sh
+
+case $(board_name) in
+        linksys,panamera)
+                # clear partialboots and always set to boot first partition
+                nvram set bootpartition=0 && nvram set partialboots=0 && nvram commit
+                ;;
+esac
```
Comment (adschm) [inline target/linux/bcm53xx/base-files/etc/init.d/clear_partialboot:12]: please use indent inside a function.

**Comment by adschm** [comment]
I would rephrase this a little to make it more clear that you are not adding an entirely new device, but improve/fix the support of an existing device (at least that's what I learned after having looked at everything eventually in the final patch). Having done that, you probably can merge the 2nd and 3rd patch together ...

File: target/linux/bcm53xx/base-files/etc/init.d/clear_partialboot (hunk 2 above)
Comment (npcomplete13) [reply target/linux/bcm53xx/base-files/etc/init.d/clear_partialboot:12]: Will do, thanks

File: target/linux/bcm53xx/base-files/etc/init.d/clear_partialboot [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+}
+
```
[the remark this answers is not in the document]
Comment (npcomplete13) [reply target/linux/bcm53xx/base-files/etc/init.d/clear_partialboot:14]: sure thing

File: target/linux/bcm53xx/base-files/etc/board.d/02_network (hunk 1 above)
Comment (npcomplete13) [reply target/linux/bcm53xx/base-files/etc/board.d/02_network:36]: I assumed these labeled the physical ports on the device, but I didn't notice them in UI. I can get rid of them.

Comment (adschm) [reply target/linux/bcm53xx/base-files/etc/board.d/02_network:36]: x:lan:y
The left number x is the internal port number of the switch, and the right number y is something corresponding to the label, if that deviates from internal order. However, I was under the impression that LuCI would not allow gaps between these numbers, but use consecutive numbering anyway ...

Comment (npcomplete13) [reply target/linux/bcm53xx/base-files/etc/board.d/02_network:36]: For now I'll just remove the ys. Thanks

File: target/linux/bcm53xx/base-files/etc/init.d/clear_partialboot (hunk 3 above)
[the remark this answers is not in the document]
Comment (npcomplete13) [reply target/linux/bcm53xx/base-files/etc/init.d/clear_partialboot]: Done

File: target/linux/bcm53xx/base-files/etc/init.d/clear_partialboot (hunk 2 above)
Comment (npcomplete13) [reply target/linux/bcm53xx/base-files/etc/init.d/clear_partialboot]: Done

File: target/linux/bcm53xx/base-files/etc/board.d/02_network (hunk 1 above)
Comment (npcomplete13) [reply target/linux/bcm53xx/base-files/etc/board.d/02_network]: Done

**Comment by npcomplete13** [comment]
> I would rephrase this a little to make it more clear that you are not adding an entirely new device, but improve/fix the support of an existing device (at least that's what I learned after having looked at everything eventually in the final patch). Having done that, you probably can merge the 2nd and 3rd patch together ...

This is done as well

**Comment by npcomplete13** [comment]
@rmilecki could you please review this patch? Thanks

**Comment by rmilecki** [comment]
What happens if someone's current boot partition is 0 and he installs OpenWrt (which AFAIU will get installed as boot partition 1)?

**Comment by npcomplete13** [comment]
> What happens if someone's current boot partition is 0 and he installs OpenWrt (which AFAIU will get installed as boot partition 1)?

There are two scenarios:

1. User is currently on factory firmware.
    - US model - needs to be flash using serial console as the firmware is signed and encrypted.
    - EU model - can be flashed using openwrt image wrapped in linksys img format. 

2. User is already on OpenWrt, then it is almost certainly on boot partition 0
    - It will remain on partition 0 and upgrade partition 0, bootflag won't be changed. partition 1 won't be touched.

**Comment by rmilecki** [comment]
OK, let's consider this case: user has UE model and runs factory firmware flashed on boot partition 0.

What happens after flashing OpenWrt (using linksys img format)?

From what I understand:
1. OpenWrt will get flashed to boot partition 1
2. Linux will believe it was flashed from boot partition 0
3. Linux/OpenWrt will try to load userspace from boot partition 0 (factory firmware)
4. Device won't boot properly

**Comment by npcomplete13** [comment]
I think the router boots fine but successive reboot will bring them back to factory.
My suggestion is that, after first boot, the user should flash OpenWrt to second partition using mtdwrite.

See here:
https://forum.openwrt.org/t/build-for-linksys-ea9500/1817/597?u=npcomplete

The other option is to flash directly using serial console.

**Comment by npcomplete13** [comment]
Hi @rmilecki 
For EU model, after testing it myself, you are right. OpenWrt loads userspace from boot partition 0 and router fails to boot OpenWrt properly and reboots to Factory.

However, you could flash to bootpartition 0 if you force reboot boot thrice using the powrt button. This forces the factory to boot from bootpartition=1.

The other route is to flash using serial cable.

**Comment by npcomplete13** [comment]
@rmilecki I have created special firmware for EU router for first time flash. This will omit the first firmware partition, hence it will no load userspace. If not (bootpartion = second partition) they can load alternate img (uses first partition) which will work as expected. A bit of hassle but works

Please approve the merge. I'll update the patch if needed. Thanks!

**Comment by npcomplete13** [comment]
Created new pull request: https://github.com/openwrt/openwrt/pull/3878

**Comment by npcomplete13** [comment]
https://github.com/openwrt/openwrt/pull/3878

**Comment by adschm** [comment]
So, what's the point of having two PRs now, and split the discussion?

**Comment by npcomplete13** [comment]
Please suggest what can be done. Deleted my previous repo and had to recreate the patch. how can I use existing PR?

**Comment by adschm** [comment]
Hi, I've finally compiled a guide for it here:
https://openwrt.org/docs/guide-developer/working-with-github-pr#reopen_closed_pr

**Comment by npcomplete13** [comment]
Reopened this pull request :)

**Comment by npcomplete13** [comment]
@adschm Can you point me what changes does this need?

I've pointed out that the installation requires serial cable. I've updated the comment in the pull request as well.

**Comment by adschm** [comment]
I do not merge bcm53xx devices, but from a quick look this still has formal issues.

Device support PR needs to contain specifications and flashing instructions, commit 3 and 4 should probably be squashed into no. 2.

**Comment by npcomplete13** [comment]
Let me fix those. Thanks!

**Comment by npcomplete13** [comment]
This is done now. Please review. Thanks!

**Comment by rmilecki** [comment]
Thank you, I'll review / handle this today

File: target/linux/bcm53xx/base-files/etc/board.d/01_leds [hunk 4]
```diff
@@ -5,6 +5,15 @@
 board_config_update
 
 case "$(board_name)" in
+linksys,panamera)
+	ucidef_set_led_usbport "usb2" "USB 2.0" "bcm53xx:green:usb2" "usb1-port2" "usb2-port2"
+	ucidef_set_led_usbport "usb3" "USB 3.0" "bcm53xx:green:usb3" "usb1-port1" "usb2-port1" "usb4-port1"
+	ucidef_set_led_netdev "wifi-enabled" "Wireless5G1" "bcm53xx:white:wifi-enabled" "wlan0" "link"
+	ucidef_set_led_default "wifi-disable-button" "Wifi-Disable" "bcm53xx:amber:wifi-disabled" "1"
+	ucidef_set_led_default "wps-disable-button" "Wifi-Disable" "bcm53xx:white:wps" "1"
+	ucidef_set_led_default "logo" "Power" "bcm53xx:white:power" "1"
```
Comment (adschm) [inline target/linux/bcm53xx/base-files/etc/board.d/01_leds:14]: Do we really need all these boolean LEDs defined here?

File: target/linux/bcm53xx/base-files/etc/board.d/01_leds [hunk 5]
```diff
@@ -5,6 +5,15 @@
 board_config_update
 
 case "$(board_name)" in
+linksys,panamera)
+	ucidef_set_led_usbport "usb2" "USB 2.0" "bcm53xx:green:usb2" "usb1-port2" "usb2-port2"
+	ucidef_set_led_usbport "usb3" "USB 3.0" "bcm53xx:green:usb3" "usb1-port1" "usb2-port1" "usb4-port1"
+	ucidef_set_led_netdev "wifi-enabled" "Wireless5G1" "bcm53xx:white:wifi-enabled" "wlan0" "link"
+	ucidef_set_led_default "wifi-disable-button" "Wifi-Disable" "bcm53xx:amber:wifi-disabled" "1"
+	ucidef_set_led_default "wps-disable-button" "Wifi-Disable" "bcm53xx:white:wps" "1"
```
Comment (adschm) [inline target/linux/bcm53xx/base-files/etc/board.d/01_leds:13]: WiFi-Disable->WPS?

File: target/linux/bcm53xx/base-files/etc/board.d/01_leds (hunk 4 above)
Comment (npcomplete13) [reply target/linux/bcm53xx/base-files/etc/board.d/01_leds:14]: It's a typo, I'll fix it

Comment (npcomplete13) [reply target/linux/bcm53xx/base-files/etc/board.d/01_leds:14]: These are necessary

File: target/linux/bcm53xx/base-files/etc/board.d/01_leds (hunk 5 above)
Comment (npcomplete13) [reply target/linux/bcm53xx/base-files/etc/board.d/01_leds:13]: I'll fix the typo

Comment (npcomplete13) [reply target/linux/bcm53xx/base-files/etc/board.d/01_leds]: fixed

File: target/linux/bcm53xx/base-files/etc/board.d/01_leds (hunk 4 above)
Comment (adschm) [reply target/linux/bcm53xx/base-files/etc/board.d/01_leds:14]: This somewhat looks like you are abusing LED definitions for setting GPIOs? Or how does a LED "wifi-disable" make sense at all?

Comment (npcomplete13) [reply target/linux/bcm53xx/base-files/etc/board.d/01_leds:14]: There is a group of eight white-led vertical bars on the router's face which will turn off when wifi is off according to the manual. These cannot be controlled individually:

image

At the moment, the MMIO leds cannot be programmed as factory. I've left them out.

Comment (adschm) [reply target/linux/bcm53xx/base-files/etc/board.d/01_leds:14]: Still, 01_leds is for defining LED _triggers_. Buttons or GPIO switches to enable anything (including LEDs) do not belong here.
Typical ways to implement that would be stuff like gpio-export, gpio hogs, or user-space code in a separate file like here:
https://github.com/openwrt/openwrt/blob/master/target/linux/ath79/generic/base-files/etc/board.d/03_gpio_switches

While I have no idea what's the proper way to handle this on bcm53xx (probably @rmilecki can help here), just abusing LED functions for buttons/switches should not be the way to go.

Comment (npcomplete13) [reply target/linux/bcm53xx/base-files/etc/board.d/01_leds:14]: I clubbed them here only because it pertains to LEDs.  If needed I can certainly  move "Wifi-Disable" and "WPS-Disable" button to separate file. I just need a confirmation from you or Rafal to do so.

Comment (npcomplete13) [reply target/linux/bcm53xx/base-files/etc/board.d/01_leds:12]: Removed them for now, I'll submit a separate pull request once I get a confirmation. I'll include proper explanation for those. Thanks

**Comment by npcomplete13** [comment]
@adschm Let me know if you want to me remove the specs from the commit. Thanks

File: target/linux/bcm53xx/base-files/etc/board.d/01_leds [hunk 6]
```diff
@@ -5,6 +5,13 @@
 board_config_update
 
 case "$(board_name)" in
+linksys,panamera)
+	ucidef_set_led_usbport "usb2" "USB 2.0" "bcm53xx:green:usb2" "usb1-port2" "usb2-port2"
+	ucidef_set_led_usbport "usb3" "USB 3.0" "bcm53xx:green:usb3" "usb1-port1" "usb2-port1" "usb4-port1"
+	ucidef_set_led_netdev "wifi-enabled" "Wireless5G1" "bcm53xx:white:wifi-enabled" "wlan0" "link"
```
Comment (adschm) [inline target/linux/bcm53xx/base-files/etc/board.d/01_leds:11]: Hmm, if this has multiple WiFi interfaces and this is only linked to one, it might be better to remove this line entirely and leave the configuration to the user.

File: target/linux/bcm53xx/base-files/etc/board.d/01_leds [hunk 7]
```diff
@@ -5,6 +5,13 @@
 board_config_update
 
 case "$(board_name)" in
+linksys,panamera)
+	ucidef_set_led_usbport "usb2" "USB 2.0" "bcm53xx:green:usb2" "usb1-port2" "usb2-port2"
+	ucidef_set_led_usbport "usb3" "USB 3.0" "bcm53xx:green:usb3" "usb1-port1" "usb2-port1" "usb4-port1"
```
Comment (rmilecki) [inline target/linux/bcm53xx/base-files/etc/board.d/01_leds:10]: You should not need `ucidef_set_led_usbport` if you describe LEDs precisely in the DT. See how it's done for other `BCM5301X` devices.

**Comment by rmilecki** [comment]
@npcomplete13: could you rebase your MR - you can drop your `bcm53xx: backport Linksys Panamera (EA9500) patches` now

On Monday I'll send upstream `mtd` patch extending fixed partitions parser. I'd like Linksys EA9500 to follow that & handle it's semi-fixed partitioning properly.

File: target/linux/bcm53xx/base-files/etc/board.d/01_leds (hunk 7 above)
Comment (npcomplete13) [reply target/linux/bcm53xx/base-files/etc/board.d/01_leds:10]: Sure thing, I have those already defined in DTS, so will drop those. Thanks.

**Comment by npcomplete13** [comment]
Will do, Thanks!

**Review by Kimamisa — APPROVED** [review]

**Comment by rmilecki** [comment]
A designed Linksys partitioning driver extension is in works, see
```
[PATCH 1/2] dt-bindings: mtd: add binding for Linksys Northstar partitions
[PATCH 2/2] mtd: parsers: ofpart: support Linksys Northstar partitions
```
https://patchwork.ozlabs.org/project/linux-mtd/list/?series=233686

This will allow reliable OpenWrt booting on Linksys EA9500.

File: target/linux/bcm53xx/base-files/etc/board.d/02_network [hunk 8]
```diff
@@ -31,6 +31,10 @@ bcm53xx_setup_interfaces()
 		ucidef_add_switch "switch0" \
 			"0:lan" "1:lan" "2:lan" "3:lan" "4:wan" "5t@eth0"
 		;;
+	linksys,panamera)
+		ucidef_add_switch "switch1" \
+			"0:lan" "1:lan" "2:lan" "3:lan" "4:wan" "5t@eth0"
```
Comment (rmilecki) [inline target/linux/bcm53xx/base-files/etc/board.d/02_network:36]: Can you use full syntax, with LAN port number, please? That should be quite helpful on this complex device (with 2 switches). I mean `n:lan:m` as for some other devices.

What about other LAN ports? In DT I can see BCM53125 handling 5 LAN ports.

Comment (npcomplete13) [reply target/linux/bcm53xx/base-files/etc/board.d/02_network:36]: I had this before, but I was told to get rid of it. I was in fact looking for an example how to use it for two switches but I could not find one. if  you can point me an example I can adapt it to my case.

`			"0:lan" "1:lan:7" "2:lan:4" "3:lan:8" "4:wan" "5t@eth0"`

As far as the other lan ports, I'm not sure the legacy switch driver can handle second switch except in dumb fwd mode, let me know if that's not the case. I've tested extensively with DSA and there is no easy way to configure DSA ports as OpenWrt lacks same framework as legacy switch driver.

File: target/linux/bcm53xx/base-files/etc/board.d/02_network (hunk 1 above)
Comment (rmilecki) [reply target/linux/bcm53xx/base-files/etc/board.d/02_network]: @adschm: that `x:lan:y` syntax works for me and I find it very useful. It results in adding `index` to switch ports in the `/etc/board.json`. UI uses that to display LAN port names, see:
switch

Comment (adschm) [reply target/linux/bcm53xx/base-files/etc/board.d/02_network]: Yes, my main point was that I don't understand why we need 7 and 8 here. Normally, the numbers for Luci would be :1 to :4 if there is four lan ports. But if it works now, it's probably not worth touching it again.

Comment (rmilecki) [reply target/linux/bcm53xx/base-files/etc/board.d/02_network]: This device has 8 LAN ports connected to 2 switches

Comment (adschm) [reply target/linux/bcm53xx/base-files/etc/board.d/02_network]: But only four are shown and numbered 1-4 in Luci, so it's technically not required. Anyway, since this preserves additional information, we might as well keep it :-)
